doc/lspconfig.txt

```diff
@@ -266,6 +266,23 @@ The global defaults for all servers can be overridden by extending the
 <
 `setup {}` can additionally override these defaults in subsequent calls.
 
+==============================================================================
+SETUP HOOK                                                *lspconfig-setup-hook*
+
+`lspconfig` will execute the `on_setup` hook for each setup call to a server after
+validating its configuration, and before attempting to launch the server
+itself. One typical usage is to allow ad-hoc substitution for any
+configuration entry, such as `cmd`.
+
+>
+  local lspconfig = require 'lspconfig'
+  lspconfig.util.on_setup = util.add_hook_before(lspconfig.util.on_setup, function(config)
+    if some_condition then
+      local custom_server_prefix = "/my/custom/path"
+      config.cmd_env = custom_server_prefix
```

Comment:
good catch!

Reply:
hopefully this new example makes sense!

> FWIW, I found use for this in my personal configs for making sure to apply a set of default configurations (primarily the ever so present `on_attach` and `capabilities` settings).

I'm not sure if this is something that @mjlbach wants to advertise (?)

since he suggested the note about this targeting plugins devs, and one can get a somewhat similar result by overriding the global defaults `:h lspconfig-global-defaults`

Reply:
Ah TIL about global defaults 😅. I guess one difference is that with the hook approach you can make sure to merge settings, for example chaining function `on_attach` and `on_new_config` calls.